### ISSUE TITLE
connlib: DNS Interception

### ISSUE THREAD

**Issue by jamilbk** [CloudFire-LLC/cloudfire#2043]
In the case where a resource `type` is `dns`, Connlib will need to intercept DNS requests in order to resolve resource names to tunnel-side IPs. The Firezone DNS "Server" could "live" at the IP `100.100.111.1`. This address could be efficiently matched on later in the data path to mangle DNS packets, replacing queries to protected resources with our static tunnel IP equivalents. For queries that don't match any Resources in the in-memory lookup table, we forward the requests to a recursive resolver set by the administrator (e.g. NextDNS) or fallback to the previously set system resolver.

DNS interception could work as follows:

1. When the VPN tunnel is started, the client app inserts a new DNS resolver for the system at highest priority to `100.100.111.1`. This is the Firezone DNS "Server". **All DNS queries will now be sent to this resolver first**.
2. In the BoringTun event loop, before the packet is encapsulated, check if `dstIp = 100.100.111.1`.
3. If so, check if it's a valid DNS request, e.g. `dstProto = (udp|tcp) AND dstPort = 53 AND type = (A|AAAA)`
4. If so, lookup the resource IP in the in-memory resource-ip table and respond with this IP.
5. If not, send request to the upstream DNS provider if one has been configured. If none configured, reply with `NXDOMAIN`.

@conectado I think the ~~[trust-dns](https://docs.rs/crate/trust-dns/latest) crate~~ could be useful here? EDIT: See https://crates.io/crates/dnssector

- [x] firezone/product#624

**Comment by jamilbk** [comment]
Aaaaaanndddd of course there's already a library geared for this: https://crates.io/crates/dnssector

cc @wiresock

**Comment by jamilbk** [comment]
@roop @conectado 

I just tested the NXDOMAIN behavior on macOS, and I can confirm that the second (or third, etc) DNS servers **ARE NOT** attempted if the first one returns NXDOMAIN. After discussing more with @conectado, we concluded it would be very strange indeed if an NXDOMAIN response caused the next DNS server to be tried.

[nxdomain-test.pcapng.zip](https://github.com/firezone/product/files/12141386/nxdomain-test.pcapng.zip)

An NXDOMAIN response after all is just as valid as any other response. @conectado confirmed in fact that `glibc` behavior is to treat NXDOMAIN as a valid response, indicating the DNS server is healthy. By default, on Linux, `glibc` will only try the next DNS server (sequentially, round robin) if the current DNS server times out. The default [timeout is 5 seconds](https://serverfault.com/questions/562079/adjusting-how-long-linux-takes-to-fail-over-to-backup-dns-server-listed-in-resol). After this time out occurs, that DNS server is marked unhealthy for some time (need to investigate), and the next DNS server is used instead _for all names_.

Unfortunately, the solution here might need to be platform specific using the platform's idiomatic Split DNS functionality. For example, on Linux resolveconf supports Split DNS: https://fedoramagazine.org/systemd-resolved-introduction-to-split-dns/#:~:text=The%20network%20interface%20is%20king,known%20as%20%E2%80%9Csplit%20DNS%E2%80%9D.

On Apple, we can assign `DNSSettings` to the `NEPacketTunnelNetworkSettings` with the `matchDomains` synced with Firezone's name-based Resources: https://developer.apple.com/documentation/networkextension/nednssettings

On Android, it doesn't look like we have native Split DNS functionality, so we'll need to fallback to the interception method described at the top of this issue already proof-of-concept'd by @conectado. Android similarly tries to limit the developer's ability to cleanly get the default system resolver, but the technique [laid out in this post on Stack Overflow](https://stackoverflow.com/questions/3070144/how-do-you-get-the-current-dns-servers-for-android) seems to be fairly robust. Namely this part (@pratikvelani -- can you comment on this approach?):

```java
   private String [] getServersMethodConnectivityManager() {
       // This code only works on LOLLIPOP and higher
       if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
           try {
               ArrayList<String> priorityServersArrayList  = new ArrayList<>();
               ArrayList<String> serversArrayList          = new ArrayList<>();
               ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(CONNECTIVITY_SERVICE);
               if (connectivityManager != null) {
                   // Iterate all networks
                   // Notice that android LOLLIPOP or higher allow iterating multiple connected networks of SAME type
                   for (Network network : connectivityManager.getAllNetworks()) {
                       NetworkInfo networkInfo = connectivityManager.getNetworkInfo(network);
                       if (networkInfo.isConnected()) {
                           LinkProperties linkProperties    = connectivityManager.getLinkProperties(network);
                           List<InetAddress> dnsServersList = linkProperties.getDnsServers();
                           // Prioritize the DNS servers for link which have a default route
                           if (linkPropertiesHasDefaultRoute(linkProperties)) {
                               for (InetAddress element: dnsServersList) {
                                   String dnsHost = element.getHostAddress();
                                   priorityServersArrayList.add(dnsHost);
                               }
                           } else {
                               for (InetAddress element: dnsServersList) {
                                   String dnsHost = element.getHostAddress();
                                   serversArrayList.add(dnsHost);
                               }
                           }
                       }
                   }
               }
               // Append secondary arrays only if priority is empty
               if (priorityServersArrayList.isEmpty()) {
                   priorityServersArrayList.addAll(serversArrayList);
               }
               // Stop here if we have at least one DNS server
               if (priorityServersArrayList.size() > 0) {
                   return priorityServersArrayList.toArray(new String[0]);
               }
           } catch (Exception ex) {
               Log.d(TAG, "Exception detecting DNS servers using ConnectivityManager method", ex);
           }
       }
       // Failure
       return null;
   }
```

Windows does have powerful DNS features and tons of backwards-compatibility so I'm sure we can find an idiomatic approach there, e.g. [Name resolution policy table](https://learn.microsoft.com/en-us/windows/security/operating-system-security/network-security/vpn/vpn-name-resolution#name-resolution-policy-table-nrpt).

So to recap, I think this may be the most robust option for each platform:

- Apple: [NEDNSSettings](https://developer.apple.com/documentation/networkextension/nednssettings)
- Windows: [Name resolution policy table](https://learn.microsoft.com/en-us/windows/security/operating-system-security/network-security/vpn/vpn-name-resolution#name-resolution-policy-table-nrpt).
- Android: Connlib DNS interception and reply/forward to default resolver using the technique above. (this seems error-prone: what if the system resolver is changed by the user while the Firezone tunnel is up? Do we cache the gathered resolver and re-retrieve it after some period of time?). Tailscale also [had issues with Android ](https://github.com/tailscale/tailscale/issues/5007)and uses a similar DNS interception approach.
- Linux: [resolved split DNS](https://fedoramagazine.org/systemd-resolved-introduction-to-split-dns/#:~:text=The%20network%20interface%20is%20king,known%20as%20%E2%80%9Csplit%20DNS%E2%80%9D.)

For the FFI changes, we could just piggyback off the `onUpdateResources` call, parse out the DNS-based Resources ("type" is `dns`) and perform the necessary Split DNS configuration updates in the client-native side. What do you think @francesca64?

  This is more work client-side, but I think the best approach given how every platform seems to be tightening down on privacy-preserving querying of the system's DNS servers.

**Comment by jamilbk** [comment]
I should clarify, all the work above is to support allowing "Fallback to System resolver" if the administrator has not specified a global DNS server for clients as fallback. Once the administrator specifies a global fallback DNS resolver, **Connlib should be switched back into DNS interception mode for _all names_, forwarding the unmatched ones to the fallback resolver**.

This is an important feature for our customers because it means a Firezone administrator can force all of the clients in his/her tenant to use a secure resolver, or DNS-over-HTTPS for example.

I believe we will need appropriate FFI modifications to support this logic:

- When the tunnel comes up, connlib receives the fallback DNS resolution strategy from the portal settings. If the fallback strategy is `system`, then:
  - For **Android**:
    - Retrieve the default system resolver using technique above
    - Enable upstream resolution in connlib, forwarding unmatched queries to default system resolver
  - For **Apple**:
    - Disable upstream resolution in connlib
    - Configure Split DNS using NEDNSSettings for the packet tunnel. Keep these settings updated upon `onUpdateResources`
  - For **Windows**: 
    - Disable upstream resolution in connlib
    - Configure Split DNS using Name resolution policy table, keeping it updated upon subsequent `onUpdateResources` calls
  - For **Linux**:
    - ~~Disable upstream resolution in connlib~~
    - ~~Configure Split DNS using resolveconf~~
    - Retrieve the default system resolver using technique above
    - Enable upstream resolution in connlib, forwarding unmatched queries to resolver
- However, if an upstream resolver is configured as the fallback strategy, do the following:
  - For **all platforms**:
    - Disable Split DNS, setting resolver to connlib's DNS_SENTINEL
    - Forward unmatched names to the upstream resolver

**Comment by jamilbk** [comment]
As a reminder, for reference, all our Wireframes are on Balsamiq (ping me if you don't have access):

![dns](https://share.balsamiq.com/c/rbRmY74D6grT5reM7HxUUT.png)

**Comment by roop** [comment]
I tested this as well (specified a DNS server that resolved one domain and returned NXDOMAIN for all other domains), set that as the DNS server for the tunnel, then checked if the OS can resolve other domains. It doesn't resolve other domains, so your call is correct. Sorry about opening up this rabbit hole.

Also agree about the idea of using split-DNS using NEDNSSettings.

**Comment by conectado** [comment]
@jamilbk

> Unfortunately, the solution here might need to be platform specific using the platform's idiomatic Split DNS functionality. For example, on Linux resolveconf supports Split DNS: https://fedoramagazine.org/systemd-resolved-introduction-to-split-dns/#:~:text=The%20network%20interface%20is%20king,known%20as%20%E2%80%9Csplit%20DNS%E2%80%9D.

Do we plan to support resources like `~.my-company.com`? otherwise we could simply use `/etc/host` I think that will be quite easier that using `systemd-resolved` which doesn't exists in every linux distro.

**Comment by conectado** [comment]
> For the FFI changes, we could just piggyback off the onUpdateResources call, parse out the DNS-based Resources ("type" is dns) and perform the necessary Split DNS configuration updates in the client-native side. What do you think @francesca64?

I like this approach, doesn't require any FFI API change.

**Comment by jamilbk** [comment]
> Do we plan to support resources like `~.my-company.com`? otherwise we could simply use `/etc/host` I think that will be quite easier that using `systemd-resolved` which doesn't exists in every linux distro.

I think `/etc/hosts` works for a Linux client MVP, but I get the feeling we'll annoy users that are using that file for other things since these are end-user machines after all.

I suppose we could make the Linux client behave the same as the Android client: read `/etc/resolv.conf` and lock connlib into upstream-forwarding mode with either the nameserver(s) found there, or the admin-configured one if it's set in the portal. That would probably work better for CI too.

**Comment by jamilbk** [comment]
Just missing:

- forwarding to upstream
- SplitDNS for Linux

**Comment by jamilbk** [comment]
Split DNS for linux will be implemented in #1762
